irohad/synchronization/synchronization.hpp

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright Soramitsu Co., Ltd. 2016 All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//
+// Created by Takumi Yamashita on 2017/04/28.
+//
+
+#ifndef IROHA_SYNCHRONIZATION_SYNCHRONIZATION_H
+#define IROHA_SYNCHRONIZATION_SYNCHRONIZATION_H
+
+#include <functional>
+#include <network/sync_client.hpp>
+#include <network/sync_gate.hpp>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include <model/model.hpp>
+
+namespace iroha {
+  namespace synchronization {
+
+    using Block = iroha::model::Block;
+
+    class BlockCompare : public std::binary_function<Block, Block, bool> {
+     public:
+      bool operator()(const Block& a, const Block& b) {
+        // NOTE: This comparison assumes we use Byzantine fault tolerance obtained through a
+        // supermajority of validating peers finalizing the block. For Nakamoto Consensus, the
+        // concept of a block score (probably based on time and difficulty or some similar metric)
+        // is needed.
+        return a.height < b.height;
+      }
+    };
+
+    // TODO
+    class Synchronizer {
+     public:
+      // When commit block after consensus, invoke this function
+      void trigger(const Block& commited_block);
+
+     private:
+      // trigger call this.
+      void startSync(const Block& commited_block);
+
+      // startSync that two thread below.
```

Comment:
This comment is somewhat confusing. You are just fetching blocks in a separate thread from the main thread, right?